lib/model/query/entities.js

```diff
@@ -89,9 +89,53 @@ const createSource = (details = null, subDefId = null, eventId = null) => ({ one
     .then((row) => row.id);
 };
 
+//
+// Entity event processing
+//
+
+const _createEntity = (dataset, entityData, submissionId, submissionDef, submissionDefId, event, parentEvent) => async ({ Entities, Audits }) => {
+  const partial = await Entity.fromParseEntityData(entityData);
+
+  const sourceDetails = { submission: { instanceId: submissionDef.instanceId }, parentEventId: parentEvent ? parentEvent.id : undefined };
+  const sourceId = await Entities.createSource(sourceDetails, submissionDefId, event.id);
+  const entity = await Entities.createNew(dataset, partial, submissionDef, sourceId);
+
+  return Audits.log({ id: event.actorId }, 'entity.create', { acteeId: dataset.acteeId },
+    {
+      entityId: entity.id, // Added in v2023.3 and backfilled
+      entityDefId: entity.aux.currentVersion.id, // Added in v2023.3 and backfilled
+      entity: { uuid: entity.uuid, dataset: dataset.name },
+      submissionId,
+      submissionDefId
+    });
+};
+
+const _updateEntity = (dataset, entityData, submissionId, submissionDef, submissionDefId, event, parentEvent) => async ({ Entities }) => {
+
+  // Get version of entity on the server
+  const serverEntity = await Entities.getById(dataset.id, entityData.system.id).then(o => o.get());
+
+  // merge data
+  const mergedData = mergeRight(serverEntity.aux.currentVersion.data, entityData.data);
+  const mergedLabel = entityData.system.label || serverEntity.aux.currentVersion.label;
+
+  // make some kind of source object
+  const sourceDetails = {
+    submission: { instanceId: submissionDef.instanceId },
+    dataParsedFromSubmission: entityData.data, // or some other info
+    labelParsedFromSubmission: entityData.label,
+    baseVersionParsedFromSubmission: 0,
+    parentEventId: parentEvent ? parentEvent.id : undefined,
```

Comment:
I don't feel like `parentEvent.id` needs to be stored here. For entity creation, changing `approvalRequired` to `false` can cause entities to be created from pending submissions. Without the change to `false`, the entity wouldn't have been created. However, that's not the case for entity updates: the `approvalRequired` flag shouldn't affect whether an entity update is completed. So I don't think we should log a change to the flag as a `parentEvent`.

Changing `approvalRequired` to `false` might happen to cause the entity update to be processed sooner than it would have otherwise, if there's a backlog in the worker queue. That's because `Entities.createEntitiesFromPendingSubmissions()` processes submission events. Yet even in that case, I don't think the change to `approvalRequired` qualifies as a `parentEvent`. The entity update would have been processed anyway, just later.

I also think we could circle back to these questions later. In that case, it'd be great if you could track this question in an issue (new or existing).

Reply:
I'll remove `parentEvent.id`. I think the rest of your comments about how `approvalRequired` plays into all this, and the order in which things get processed, is one of the themes behind https://github.com/getodk/central/issues/517 so we can come back to it then.